leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/RegistrationTest.java

```diff
@@ -86,6 +90,38 @@ public void register_update_deregister() {
     }
 
     @Test
+    public void deregister_cancel_pending_request() throws InterruptedException {
+        // Check registration
+        assertTrue(helper.server.getClientRegistry().allClients().isEmpty());
+
+        helper.client.start();
+        helper.waitForRegistration(1);
+
+        assertEquals(1, helper.server.getClientRegistry().allClients().size());
+        Client client = helper.server.getClientRegistry().get(ENDPOINT_IDENTIFIER);
+        assertNotNull(client);
+        assertArrayEquals(LinkObject.parse("</>;rt=\"oma.lwm2m\",</1/0>,</2>,</3/0>".getBytes()),
+                client.getObjectLinks());
+
+        // Stop client with out de-registration
+        helper.client.stop(false);
+
+        // Send a read which should be retransmitted.
+        Callback<ReadResponse> callback = new Callback<ReadResponse>();
+        helper.server.send(client, new ReadRequest(3, 0, 1), callback, callback);
+
+        // Restart client (de-registration/re-registration)
+        helper.client.start();
+
+        // Check the request was cancelled.
+        boolean timedout = !callback.waitForResponse(1000);
+        assertFalse("Response or Error expected", timedout);
+        assertTrue("Response or Error expected", callback.isCalled().get());
+        assertNull("No response expected", callback.getResponse());
+        assertNotNull("Exception expected", callback.getException());
+    }
+
```

Comment:
@Test
    public void deregister_cancel_multiple_pending_request() throws InterruptedException {
        // Check registration
        assertTrue(helper.server.getClientRegistry().allClients().isEmpty());

        helper.client.start();
        helper.waitForRegistration(1);

        assertEquals(1, helper.server.getClientRegistry().allClients().size());
        Client client = helper.server.getClientRegistry().get(ENDPOINT_IDENTIFIER);
        assertNotNull(client);
        assertArrayEquals(LinkObject.parse("</>;rt=\"oma.lwm2m\",</1/0>,</2>,</3/0>".getBytes()),
                client.getObjectLinks());

        // Stop client with out de-registration
        helper.client.stop(false);

        // Send multiple reads which should be retransmitted.
        List<Callback<ReadResponse>> callbacks = new ArrayList<Callback<ReadResponse>>();

        for (int index = 0; index < 4; ++index) {
            Callback<ReadResponse> callback = new Callback<ReadResponse>();
            helper.server.send(client, new ReadRequest(3, 0, 1), callback, callback);
            callbacks.add(callback);
        }

        // Restart client (de-registration/re-registration)
        helper.client.start();

        // Check the request was cancelled.
        int index = 0;
        for (Callback<ReadResponse> callback : callbacks) {
            boolean timedout = !callback.waitForResponse(1000);
            assertFalse("Response or Error expected, no timeout, call " + index, timedout);
            assertTrue("Response or Error expected, call " + index, callback.isCalled().get());
            assertNull("No response expected, call " + index, callback.getResponse());
            assertNotNull("Exception expected, call " + index, callback.getException());
            ++index;
        }
    }

Debugging this test shows:
- remove during iterate throws an exception
- that exception is "silently ignored"